### PR TITLE
Make String(unsafeUninitializedCapacity:initializingUTF8With:) public

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Review by milseman — APPROVED** [review]

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by benrimmington** [comment]
Should the public initializer include the `unsafe` label from the [review](https://forums.swift.org/t/se-0263-review-2-add-a-string-initializer-with-access-to-uninitialized-storage/29663)?

In which case, would the internal initializer become legacy ABI?

**Comment by Catfish-Man** [comment]
Oh did I get that backwards? For some reason I thought we'd ended up removing it. Thanks for catching that!

Also what is swift-ci doing :(

**Comment by benrimmington** [comment]
[Swift CI Outage - Feb 28th, 2020 @ 10:00 AM PST](https://forums.swift.org/t/swift-ci-outage-feb-28th-2020-10-00-am-pst/34150)

File: stdlib/public/core/String.swift [hunk 1]
```diff
@@ -476,7 +476,8 @@ extension String {
   ///       - buffer: A buffer covering uninitialized memory with room for the
   ///           specified number of UTF-8 code units.
   @inline(__always)
-  internal init(
+  @available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+  public init(
```
Comment (milseman) [inline stdlib/public/core/String.swift]:
```
  public init(
    unsafeUninitializedCapacity capacity: Int,
    initializingUTF8With initializer: (
      _ buffer: UnsafeMutableBufferPointer<UInt8>,
    ) throws -> Int
  ) rethrows
```

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by benrimmington** [comment]
The `unsafeUninitializedCapacity:` label also needs to be used in existing callers:

* [`lowercased()`](https://github.com/apple/swift/blob/e615c1ca16ad75e75d4c823e206c393277b87bb8/stdlib/public/core/String.swift#L789)

* [`uppercased()`](https://github.com/apple/swift/blob/e615c1ca16ad75e75d4c823e206c393277b87bb8/stdlib/public/core/String.swift#L849)

* [`_foreignGrow(_:)`](https://github.com/apple/swift/blob/e615c1ca16ad75e75d4c823e206c393277b87bb8/stdlib/public/core/StringGutsRangeReplaceable.swift#L100)

* [`_foreignConvertedToSmall()`](https://github.com/apple/swift/blob/e615c1ca16ad75e75d4c823e206c393277b87bb8/stdlib/public/core/StringGutsRangeReplaceable.swift#L164)

**Comment by Catfish-Man** [comment]
Huh when did lowercased() and friends start using this, that's neat

**Comment by Catfish-Man** [comment]
Also I have to work around availability >.<

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by benrimmington** [comment]
The public and internal initializers are `@inline(__always)` but not `@inlinable`. Does that mean they are inlined within stdlib, but not inlined into other modules?

And if the callers I listed above (`lowercased()`, etc.) are also not inlined into other modules, then shouldn't you be able to use the public initializer everywhere, and remove the internal initializer completely?

The "test and merge" failed.

> <https://ci.swift.org/job/swift-PR-merge/2268/>
> ERROR: Couldn't find any revision to build.

You could try a "test" only, and merge manually if that works.

**Comment by Catfish-Man** [comment]
Yes, inline always but not inlinable was intentional here. I want to try taking that off at some point, but getting the SmallString part inlined as much as possible is important (originally I wanted to make it inlinable, but SmallString isn't ABI).

I tried using the public initializer in lowercased() and friends but it still got upset about availability :(

test only is a good idea, let's try that

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by benrimmington** [comment]
All checks have passed!

**Pull request merged**